src/providers/mod.rs

```diff
@@ -254,3 +284,67 @@ pub trait MetadataProvider {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tempfile::NamedTempFile;
+
+    struct HostnameMock(String);
+
+    impl MetadataProvider for HostnameMock {
+        fn hostname(&self) -> Result<Option<String>> {
+            Ok(Some(self.0.clone()))
+        }
+    }
+
+    // write specified hostname to a file, then read it back
+    fn try_write_hostname(hostname: &str) -> String {
+        let mut temp = NamedTempFile::new().unwrap();
+        let provider = HostnameMock(hostname.into());
+        provider
+            .write_hostname(temp.path().to_str().unwrap().into())
+            .unwrap();
+        let mut ret = String::new();
+        temp.read_to_string(&mut ret).unwrap();
+        ret.trim_end().into()
+    }
+
+    #[test]
+    fn test_hostname_truncation() {
```

Comment:
Maybe add the test case from https://bugzilla.redhat.com/show_bug.cgi?id=2008521 here i.e.
`ci-op-dgnrb143-44266-42t6h-worker-c-ldlk9.com.openshift-gce-devel-ci.internal`
explicitly?

Reply:
It's possibly too paranoid, but I'm using all synthetic test cases because I didn't mock the value of `HOST_NAME_MAX` and wanted to be robust to other values.